### PR TITLE
Transactional resize fixes

### PR DESCRIPTION
**Pull request by kennylevinsen**
Closes: https://github.com/swaywm/sway/issues/5848

### CODE REVIEW

**Comment by kennylevinsen** [comment]
cc @RPigott @emersion

**Comment by Xyene** [comment]
Thanks for working on this!

It looks like something's not quite right with titlebars during interactive resizes, though (excuse the vertical monitor):


https://user-images.githubusercontent.com/1403503/106537327-15fbea80-64c8-11eb-81b7-e4103ccd2e84.mp4

**Comment by kennylevinsen** [comment]
@Xyene Hmm, something seems to have been broken in the refactor when I tried to re-enable the centering logic. I'll take a look. Thanks for the test.

**Comment by kennylevinsen** [comment]
Hmm.

I think we're dealing with a bad interaction with the time of sampling and saved buffers.

The initial implementation added an extra geometry field to `struct sway_view`. Every surface commit would update `next_geometry`, with an executing transaction copying to `current_geometry` - *not* state from when the transaction was queued. This interacted poorly with surface centering logic, but otherwise *appeared* to behave correctly.

The current approach uses `struct sway_container_state`, and all the usual logic there. This means that the next geometry is copied when the transaction is queued, and moved to current when it executes. This causes occasional misrendering.

I suspect this is an interaction between when the geometry is sampled vs. e.g. when the buffer is saved. It's possible that this can be fixed by simply sampling at the time of queuing rather than when the transaction executes, not sure. The transaction system is a bit cumbersome to deal with...

**Comment by kennylevinsen** [comment]
I have reworked the approach.

It turns out most of the issues were associated with the view centering logic, especially when combined with saved buffers, and I had just been lead down a wrong path with respect to double-buffering view geometries.

I'll try to daily drive this a bit, but @RPigott and @Xyene, if you have time, please do test and see how things work out. Especially if you have any weird clients, like the GNOME polkit dialog.

**Comment by kennylevinsen** [comment]
As pointed out in IRC, horizontal resizes are now fine, but vertical resizes when the top of a window is moved still has some slack left proportional to the resize. Funnily enough, it does not happen when resizing from the bottom.

File: include/sway/tree/view.h [hunk 1]
```diff
@@ -317,6 +317,7 @@ void view_map(struct sway_view *view, struct wlr_surface *wlr_surface,
 void view_unmap(struct sway_view *view);
 
 void view_update_size(struct sway_view *view, int width, int height);
+void view_center(struct sway_view *view);
```
Comment (emersion) [inline include/sway/tree/view.h:320]: Maybe rename to `view_center_surface`? `view_center` sounds like the function would center the view on screen.

Comment (kennylevinsen) [reply include/sway/tree/view.h]: Agreed

**Comment by kennylevinsen** [comment]
Found the issue - view_center, now view_center_surface, was erroneously referring to `con->content_y` (the pending container state), rather than `con->current.content_y` (the current container state) as intended. This was throwing off surface_y, which meant that resizes that moved a surface vertically would still have glitches.

File: sway/desktop/xdg_shell.c [hunk 2]
```diff
@@ -287,10 +287,8 @@ static void handle_commit(struct wl_listener *listener, void *data) {
 
 	struct wlr_box new_geo;
 	wlr_xdg_surface_get_geometry(xdg_surface, &new_geo);
-	bool new_size = new_geo.width != view->geometry.width ||
-			new_geo.height != view->geometry.height ||
-			new_geo.x != view->geometry.x ||
-			new_geo.y != view->geometry.y;
+	bool new_size = memcmp(&new_geo, &view->geometry,
```
Comment (emersion) [inline sway/desktop/xdg_shell.c:290]: Hm, I'm concerned about padding. The struct _probably_ won't have padding, but `memcmp` would make it extremely difficult to debug if it ever has. Maybe introduce a helper for this?

Comment (kennylevinsen) [reply sway/desktop/xdg_shell.c]: Ugh, you're right, undid the change. :|

**Review by emersion — APPROVED** [review]
LGTM. @RPigott, @Xyene, do you have anything to add?

**Comment by RPigott** [comment]
Hey, sorry I'm late.

The approach makes sense. I've just tried to break it for a while with a variety of weird clients but it has held up :+1:. At the very least, the previously known issues addressed and/or caused by #5812 seem to be working. I only noticed that #5857 is still here, but I see you've already made another PR for that. I'll keep running this, but I think it could be merged.

Thanks very much for helping fix this.

**Comment by Xyene** [comment]
Thanks again for working on this! The vertical resizes are much better than before, but there seems to be at least one more case where the "sliver of the desktop" appears when resizing containers with titlebars.

Consider this layout: `V[A T[H[B* C]]]`. Resizing vertically causes a gap between B/C's titlebars and the titlebar of the `T` to appear. No gap appears between the titlebar of B/C and their content, though, so that's fixed :)

**Comment by emersion** [comment]
Alight, merging now then, thanks!

**Comment by kennylevinsen** [comment]
@Xyene I can't seem to reproduce the last mentioned problem with a `V[foot T[H[slow-ack-configure foot]]]`. Could you create a new issue for it with more details?

**Pull request merged**